chalice/local/internal.py

```diff
@@ -0,0 +1,126 @@
+# -*- coding: utf-8 -*-
+"""
+This is intended only for `pytest-chalice`.
+
+https://github.com/studio3104/pytest-chalice
+"""
+
+import json
+import re
+
+from logging import getLogger
+from typing import Any, Callable, Dict, Optional  # noqa
+
+from chalice import Chalice  # noqa
+from chalice.config import Config
+from chalice.local import LocalGateway
+from chalice.local import EventType, HeaderType  # noqa
+
+logger = getLogger(__name__)
+
+# Python 3.4 or older don't have JSONDecodeError within json module
+try:
+    JSONDecodeError = json.JSONDecodeError  # type: ignore
+except AttributeError:
+    JSONDecodeError = ValueError  # type: ignore
+
+
+class InternalLocalGateway(LocalGateway):
+    def __init__(self, *args, **kwargs):
+        # type: (Any, Any) -> None
+        self.__custom_context = {}  # type: Dict
+        super(InternalLocalGateway, self).__init__(*args, **kwargs)
+
+    @property
+    def custom_context(self):
+        # type: () -> Dict[str, Any]
+        return self.__custom_context
+
+    @custom_context.setter
+    def custom_context(self, context):
+        # type: (Dict[str, Any]) -> None
+        self.__custom_context = context
+
+    def _generate_lambda_event(self, method, path, headers, body):
+        # type: (str, str, HeaderType, Optional[str])-> EventType
+        event = super(
+            InternalLocalGateway, self)._generate_lambda_event(
+                method, path, headers, body)
+        event['requestContext'].update(self.custom_context)
+        return event
+
+
+UPPERCASE_PATTERN = re.compile('([A-Z])')
+
+
+class ResponseHandler(object):
+    def __init__(self, values):
+        # type: (Dict[str, Any]) -> None
+        self.values = {}  # type: Dict
+
+        for key, value in values.items():
+            snake_key = re.sub(
+                UPPERCASE_PATTERN, lambda x: '_' + x.group(1).lower(), key)
+            self.values[snake_key] = value
+
+        try:
+            self.values['json'] = json.loads(self.values['body'])
+        except JSONDecodeError:  # type: ignore
+            logger.info(
+                'Response body is NOT JSON decodable: %s',
+                self.values['body'])
+
+    def __getattr__(self, key):
+        # type: (str) -> Any
+        try:
+            return self.values[key]
+        except KeyError:
+            raise AttributeError(
+                "'%s' object has no attribute '%s'" % (
+                    self.__class__.__name__, key))
+
+
+class RequestHandler(object):
+    METHODS = (
+        'get', 'head', 'post', 'options', 'put',
+        'delete', 'trace', 'patch', 'link', 'unlink')
+
+    def __init__(self, app):
+        # type: (Chalice) -> None
+        self.local_gateway = InternalLocalGateway(app, Config())
+
+    @property
+    def custom_context(self):
```

Comment:
Instead of having this `custom_context` property, what if we just make that a parameter to the client?

```
client.get('/', custom_context={...})
```

Reply:
`custom_context` is not a request parameter but an object to override the response, so I worried about how I should implement this interface.
And, I thought that I wanted to make this an object member of the client instance because I just wanted to write tests just like following,

```python
@pytest.fixture
def admin_client(client):
    client.custom_context = {...}  # e.g.) dummy Cognito's context for an user which belongs to `admin` group
    return client


def test_admin_behavior(admin_client):
    response = admin_client.get(...)
    # Assetions...
```

I guess this allows me to write the test simply, but how do you think??

Reply:
Ref: https://github.com/aws/chalice/pull/1193#discussion_r352496402

Reply:
Ref: https://github.com/aws/chalice/pull/1193#discussion_r352997311